src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,691 @@
+                            result.insert(OutputPool::Sapling);
+                        }
+                        break;
+                    case ReceiverType::Orchard:
+                        if (params.GetConsensus().NetworkUpgradeActive(anchorHeight, Consensus::UPGRADE_NU5)
+                            && (!spendable.orchardNoteMetadata.empty() || strategy.AllowRevealedAmounts())) {
+                            result.insert(OutputPool::Orchard);
+                        }
+                        break;
+                }
+            }
+            return result;
+        };
+
+        changeAddr = std::visit(match {
+            [&](const CKeyID& keyId) -> ChangeAddress {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        sendFromAccount,
+                        getAllowedChangePools({ReceiverType::P2PKH}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                return sendTo.value();
+            },
+            [&](const CScriptID& scriptId) -> ChangeAddress {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        sendFromAccount,
+                        getAllowedChangePools({ReceiverType::P2SH}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                return sendTo.value();
+            },
+            [](const libzcash::SproutPaymentAddress& addr) -> ChangeAddress {
+                // for Sprout, we return change to the originating address using the tx builder.
+                return addr;
+            },
+            [](const libzcash::SproutViewingKey& vk) -> ChangeAddress {
+                // for Sprout, we return change to the originating address using the tx builder.
+                return vk.address();
+            },
+            [&](const libzcash::SaplingPaymentAddress& addr) -> ChangeAddress {
+                // for Sapling, if using a legacy address, return change to the
+                // originating address; otherwise return it to the Sapling internal
+                // address corresponding to the UFVK.
+                std::optional<RecipientAddress> sendTo = sendFromAccount == ZCASH_LEGACY_ACCOUNT
+                    ? addr
+                    : pwalletMain->GenerateChangeAddressForAccount(
+                            sendFromAccount,
+                            getAllowedChangePools({ReceiverType::Sapling}));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                return sendTo.value();
```

Comment:
If I'm reading this correctly, we are creating explicit change outputs in these cases. What then is the purpose of saving the change address and providing it to the builder later?

Alternatively, if `Builder::SendChangeTo` is sufficient, why are we creating explicit change outputs here? Is it to enable the transaction effects can be described fully? If that's the case, then it feels like `Builder::SendChangeTo` has been made obsolete and should be removed (instead requiring the caller to fully specify all outputs including change).

Reply:
There is a separate PR that does remove `SendChangeTo`, but it’s after the `z_shieldcoinbase` PR[^1], because currently `z_shieldcoinbase` treats its single output _as_ change.

I’m not sure why we set the changeAddr there. @nuttycom, any idea, or can I safely remove that code?

[^1]: The PR is currently after the `mergetoaddress` change, but I realize that `SendChangeTo` isn’t involved there, so I just moved it up in the dependency graph.

Reply:
A bit more detail – we add it as an explicit output _unless_ it’s Sprout change. We currently don’t support Sprout in `RecipientAddress`, which is the variant used by `WalletTxBuilder`. So, I think we’ll need to continue calling that for Sprout until the follow-up PR, but can perhaps remove the non-Sprout case now.

Reply:
The intent is for the `PreparedTransaction` to be the fully prepared tx, with all outputs *including* the change output; that way, any pool-crossing behavior related to change is also part of the proposal and can have its privacy impacts analyzed.

Reply:
There should be no additional change address determination after the proposal is constructed. This is also partially looking forward to the finding-the-fixed-point thing that the ZIP 317 implementation will need to do.

Reply:
My original intent had been to remove SendChangeTo as part of this PR; the decision to split out `z_shieldcoinbase` updates was made later.

Reply:
Even if `SendChangeTo` isn't removed as part of this PR, the new transaction flow should make no use of it.

Reply:
Ok, so now I’ve removed the `SendChangeTo` call, but left `SendChangeToSprout` in place.

The later PR changes `ResolvedPayments` (which is now understood to be internal, unlike `Payments`) to be like `ChangeAddress` (which allows Sprout). That PR does leave SendChangeToSprout in existence for now, because there is an RPC call (`sapling_migration`) that relies on it, but that too could be changed in another separate PR.